File: backend/app/services/passage_service.py
import re
from typing import List, Optional
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.passage import Passage, PassageCategory, PassageDifficulty, PassageLanguage
from uuid import UUID


class PassageService:
    async def get_passage(self, db: AsyncSession, passage_id: UUID) -> Optional[Passage]:
        result = await db.execute(select(Passage).where(Passage.id == passage_id, Passage.is_active == True))
        return result.scalar_one_or_none()

    async def get_passages(
        self,
        db: AsyncSession,
        category: Optional[PassageCategory] = None,
        difficulty: Optional[PassageDifficulty] = None,
        language: Optional[PassageLanguage] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Passage]:
        query = select(Passage).where(Passage.is_active == True)

        if category:
            query = query.where(Passage.category == category)
        if difficulty:
            query = query.where(Passage.difficulty == difficulty)
        if language:
            query = query.where(Passage.language == language)

        query = query.order_by(Passage.created_at.desc()).offset(offset).limit(limit)
        result = await db.execute(query)
        return result.scalars().all()

    async def get_random_passage(
        self,
        db: AsyncSession,
        category: PassageCategory = PassageCategory.SSC_CHSL,
        difficulty: PassageDifficulty = PassageDifficulty.MEDIUM,
    ) -> Optional[Passage]:
        is_ssc = category in (PassageCategory.SSC_CHSL, PassageCategory.SSC_CGL)

        if is_ssc:
            query = select(Passage).where(
                Passage.is_active == True,
                Passage.category == category,
                Passage.is_exam_length == True,
            )
            if difficulty:
                query = query.where(Passage.difficulty == difficulty)
            query = query.order_by(func.random()).limit(1)
            result = await db.execute(query)
            passage = result.scalar_one_or_none()
            if passage:
                return passage

        query = select(Passage).where(
            Passage.is_active == True,
            Passage.category == category,
        )
        if difficulty:
            query = query.where(Passage.difficulty == difficulty)
        query = query.order_by(func.random()).limit(1)
        result = await db.execute(query)
        return result.scalar_one_or_none()

    def _format_content(self, content: str, language: Optional[str] = None) -> str:
        if not content:
            return content
        if language == "hindi":
            return content.strip()

        text = content.strip()
        text = re.sub(r'  +', ' ', text)

        words = text.split()
        title_case_words = sum(1 for w in words if w and w[0].isupper() and len(w) > 1)
        total_words = len(words)

        if total_words >= 5 and title_case_words > total_words * 0.8:
            sentence_lowered = []
            for i, w in enumerate(words):
                if i == 0:
                    sentence_lowered.append(w[0].upper() + w[1:].lower() if len(w) > 1 else w.upper())
                else:
                    sentence_lowered.append(w.lower())
            text = ' '.join(sentence_lowered)

        if text and text[0].islower():
            text = text[0].upper() + text[1:]

        sentences = re.split(r'(?<=[.!?])\s+', text)
        formatted = []
        for s in sentences:
            s = s.strip()
            if s and not s[-1] in '.!?':
                s += '.'
            formatted.append(s)
        text = ' '.join(formatted)

        return text

    async def create_passage(self, db: AsyncSession, data: dict) -> Passage:
        for field in ("content", "content_hindi", "title"):
            if data.get(field):
                cleaned = re.sub(r'<[^>]*>', '', str(data[field]))
                cleaned = re.sub(r'[\x00-\x08\x0b\x0c\x0e-\x1f]', '', cleaned)
                cleaned = cleaned.replace("javascript:", "").replace("data:", "").replace("vbscript:", "")
                data[field] = cleaned[:10000]

        if data.get("content"):
            data["content"] = self._format_content(data["content"], data.get("language"))
        if data.get("content_hindi"):
            data["content_hindi"] = self._format_content(data["content_hindi"], "hindi")

        passage = Passage(**data)
        db.add(passage)
        await db.flush()
        await db.refresh(passage)
        return passage

    async def increment_usage(self, db: AsyncSession, passage_id: UUID):
        passage = await self.get_passage(db, passage_id)
        if passage:
            passage.times_used += 1
            await db.flush()


passage_service = PassageService()

File: frontend/src/server/services/passage-service.ts
import { db } from '../db/client';
import { passages } from '../db/schema/passages';
import { eq, and, desc, sql, count } from 'drizzle-orm';

export class PassageService {
  async getPassage(passageId: string) {
    const [passage] = await db
      .select()
      .from(passages)
      .where(and(eq(passages.id, passageId), eq(passages.isActive, true)))
      .limit(1);
    return passage ?? null;
  }

  async getPassages(params: {
    category?: string;
    difficulty?: string;
    language?: string;
    limit?: number;
    offset?: number;
  }) {
    const conditions = [eq(passages.isActive, true)];
    if (params.category) conditions.push(eq(passages.category, params.category));
    if (params.difficulty) conditions.push(eq(passages.difficulty, params.difficulty));
    if (params.language) conditions.push(eq(passages.language, params.language));

    return db
      .select()
      .from(passages)
      .where(and(...conditions))
      .orderBy(desc(passages.createdAt))
      .offset(params.offset ?? 0)
      .limit(params.limit ?? 50);
  }

  async getRandomPassage(params: {
    category?: string;
    difficulty?: string;
    practiceSet?: number;
  }) {
    const category = params.category ?? 'ssc_chsl';
    const difficulty = params.difficulty ?? 'medium';
    const practiceSet = params.practiceSet;
    const isSsc = ['ssc_chsl', 'ssc_cgl'].includes(category);

    if (isSsc && !practiceSet) {
      const conditions = [
        eq(passages.isActive, true),
        eq(passages.category, category),
        eq(passages.isExamLength, true),
      ];
      if (difficulty) conditions.push(eq(passages.difficulty, difficulty));

      const [{ total }] = await db
        .select({ total: count() })
        .from(passages)
        .where(and(...conditions));

      if (total > 0) {
        const offset = Math.floor(Math.random() * total);
        const [passage] = await db
          .select()
          .from(passages)
          .where(and(...conditions))
          .offset(offset)
          .limit(1);
        if (passage) return passage;
      }
    }

    const conditions = [
      eq(passages.isActive, true),
      eq(passages.category, category),
    ];
    if (difficulty) conditions.push(eq(passages.difficulty, difficulty));
    if (practiceSet) conditions.push(eq(passages.practiceSet, practiceSet));

    const [{ total }] = await db
      .select({ total: count() })
      .from(passages)
      .where(and(...conditions));

    if (total === 0) return null;

    const offset = Math.floor(Math.random() * total);
    const [passage] = await db
      .select()
      .from(passages)
      .where(and(...conditions))
      .offset(offset)
      .limit(1);

    return passage ?? null;
  }

  formatContent(content: string, language?: string): string {
    if (!content) return content;
    if (language === 'hindi') return content.trim();

    let text = content.trim();
    text = text.replace(/  +/g, ' ');

    const words = text.split(/\s+/);
    const titleCaseWords = words.filter(w => w && w[0] >= 'A' && w[0] <= 'Z' && w.length > 1).length;
    const totalWords = words.length;

    if (totalWords >= 5 && titleCaseWords > totalWords * 0.8) {
      text = words.map((w, i) => {
        if (i === 0) return w.length > 1 ? w[0].toUpperCase() + w.slice(1).toLowerCase() : w.toUpperCase();
        return w.toLowerCase();
      }).join(' ');
    }

    if (text.length > 0 && text[0] >= 'a' && text[0] <= 'z') {
      text = text[0].toUpperCase() + text.slice(1);
    }

    const sentences = text.split(/(?<=[.!?])\s+/);
    text = sentences.map(s => {
      s = s.trim();
      if (s.length > 0 && !'.!?'.includes(s[s.length - 1])) {
        s += '.';
      }
      return s;
    }).join(' ');

    return text;
  }

  async createPassage(data: Record<string, any>) {
    const cleaned: Record<string, any> = {};
    for (const field of ['content', 'content_hindi', 'title']) {
      if (data[field]) {
        let val = String(data[field])
          .replace(/<[^>]*>/g, '')
          .replace(/[\x00-\x08\x0b\x0c\x0e-\x1f]/g, '')
          .replace(/javascript:/g, '')
          .replace(/data:/g, '')
          .replace(/vbscript:/g, '')
          .slice(0, 10000);
        cleaned[field] = val;
      }
    }

    const insertData = {
      title: cleaned.title ?? data.title,
      content: cleaned.content ?? data.content,
      contentHindi: cleaned.content_hindi ?? data.content_hindi,
      language: data.language ?? 'english',
      category: data.category,
      difficulty: data.difficulty ?? 'medium',
      exactKeyDepressions: data.exactKeyDepressions,
      wordCount: data.wordCount,
      topic: data.topic,
      source: data.source,
      sscExamYear: data.sscExamYear,
      isExamLength: data.isExamLength ?? false,
    };

    if (insertData.content) {
      insertData.content = this.formatContent(insertData.content, insertData.language);
    }

    const [passage] = await db.insert(passages).values(insertData).returning();
    return passage;
  }

  async incrementUsage(passageId: string) {
    const passage = await this.getPassage(passageId);
    if (passage) {
      await db
        .update(passages)
        .set({ timesUsed: (passage.timesUsed ?? 0) + 1 })
        .where(eq(passages.id, passageId));
    }
  }
}

export const passageService = new PassageService();
